File: backend/app/api/v1/endpoints/conversations.py
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException, status
from supabase import Client
from datetime import datetime

from app.core.dependencies import get_current_active_user
from app.services.supabase_client import get_supabase
from app.api.v1.schemas.conversation import (
    ConversationResponse,
    ConversationCreate,
    AddMessageRequest,
    ConversationWithItinerary,
)
from app.config.logging import get_logger

router = APIRouter()
logger = get_logger(__name__)


@router.get("/active", response_model=Optional[ConversationWithItinerary])
async def get_active_conversation(
    current_user: dict = Depends(get_current_active_user),
    supabase: Client = Depends(get_supabase)
):
    """
    Get user's active conversation with latest itinerary.
    Returns null if no active conversation exists.
    """
    try:
        # Use the helper function from migration
        response = supabase.rpc(
            'get_active_conversation',
            {'p_user_id': current_user['id']}
        ).execute()
        
        if not response.data or len(response.data) == 0:
            return None
            
        return response.data[0]
        
    except Exception as e:
        logger.error(f"Error fetching active conversation: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to fetch active conversation"
        )


@router.post("", response_model=ConversationResponse, status_code=status.HTTP_201_CREATED)
async def create_conversation(
    conversation: ConversationCreate,
    current_user: dict = Depends(get_current_active_user),
    supabase: Client = Depends(get_supabase)
):
    """
    Create a new conversation.
    Automatically deactivates any existing active conversation.
    """
    try:
        # First, deactivate any existing active conversations
        supabase.table("conversations")\
            .update({"is_active": False})\
            .eq("user_id", current_user["id"])\
            .eq("is_active", True)\
            .execute()
        
        # Create new conversation
        data = {
            "user_id": current_user["id"],
            "messages": [msg.model_dump() for msg in conversation.messages],
            "state": conversation.state,
            "is_active": True,
            "last_message_at": datetime.utcnow().isoformat(),
        }
        
        response = supabase.table("conversations").insert(data).execute()
        
        if not response.data:
            raise ValueError("Supabase insert returned no data")
        
        return response.data[0]
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error creating conversation: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to create conversation: {str(e)}"
        )


@router.patch("/{conversation_id}/messages", response_model=ConversationResponse)
async def add_message_to_conversation(
    conversation_id: str,
    request: AddMessageRequest,
    current_user: dict = Depends(get_current_active_user),
    supabase: Client = Depends(get_supabase)
):
    """
    Add a message to an existing conversation.
    Updates last_message_at automatically via trigger.
    """
    try:
        # First, get the current conversation
        conv_response = supabase.table("conversations")\
            .select("*")\
            .eq("id", conversation_id)\
            .eq("user_id", current_user["id"])\
            .single()\
            .execute()
        
        if not conv_response.data:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Conversation not found"
            )
        
        current_messages = conv_response.data.get("messages", [])
        
        # Append new message
        new_message = request.message.model_dump()
        updated_messages = current_messages + [new_message]
        
        # Update conversation
        update_response = supabase.table("conversations")\
            .update({
                "messages": updated_messages,
                "last_message_at": datetime.utcnow().isoformat(),
            })\
            .eq("id", conversation_id)\
            .eq("user_id", current_user["id"])\
            .execute()
        
        if not update_response.data:
            raise ValueError("Failed to update conversation")
        
        return update_response.data[0]
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error adding message to conversation: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to add message: {str(e)}"
        )


@router.patch("/{conversation_id}/deactivate", status_code=status.HTTP_204_NO_CONTENT)
async def deactivate_conversation(
    conversation_id: str,
    current_user: dict = Depends(get_current_active_user),
    supabase: Client = Depends(get_supabase)
):
    """
    Deactivate a conversation (mark as inactive).
    Used when starting a new conversation.
    """
    try:
        supabase.table("conversations")\
            .update({"is_active": False})\
            .eq("id", conversation_id)\
            .eq("user_id", current_user["id"])\
            .execute()
        
        return None
        
    except Exception as e:
        logger.error(f"Error deactivating conversation: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to deactivate conversation"
        )


@router.get("/{conversation_id}", response_model=ConversationResponse)
async def get_conversation(
    conversation_id: str,
    current_user: dict = Depends(get_current_active_user),
    supabase: Client = Depends(get_supabase)
):
    """Get a specific conversation by ID."""
    try:
        response = supabase.table("conversations")\
            .select("*")\
            .eq("id", conversation_id)\
            .eq("user_id", current_user["id"])\
            .single()\
            .execute()
        
        if not response.data:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Conversation not found"
            )
        
        return response.data
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error fetching conversation: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to fetch conversation"
        )

File: frontend/src/lib/api/conversations.ts
import { createClientComponentClient } from '@supabase/auth-helpers-nextjs';

// ── Types ─────────────────────────────────────────────────────────────────────

export interface Message {
    role: 'user' | 'assistant';
    content: string;
    timestamp?: string;
    metadata?: Record<string, any>;
}

export interface Conversation {
    id: string;
    user_id: string;
    title?: string;
    is_active: boolean;
    messages: Message[];
    last_message_at: string;
    created_at: string;
    updated_at: string;
}

export interface ConversationWithItinerary extends Conversation {
    latest_itinerary?: any;
}

// ── Supabase Conversations API ────────────────────────────────────────────────

const supabase = createClientComponentClient();

export const conversationsApi = {
    /** Get the currently active conversation for the logged-in user */
    getActive: async (): Promise<ConversationWithItinerary | null> => {
        const { data: { user } } = await supabase.auth.getUser();
        if (!user) return null;

        const { data, error } = await supabase
            .from('conversations')
            .select('*')
            .eq('user_id', user.id)
            .eq('is_active', true)
            .order('updated_at', { ascending: false })
            .limit(1)
            .maybeSingle();

        if (error) throw error;
        return data;
    },

    /** Create a new conversation, deactivating any previous active one */
    create: async (): Promise<Conversation> => {
        const { data: { user } } = await supabase.auth.getUser();
        if (!user) throw new Error('Not authenticated');

        // Deactivate any currently active conversations
        await supabase
            .from('conversations')
            .update({ is_active: false })
            .eq('user_id', user.id)
            .eq('is_active', true);

        // Create new active conversation
        const { data, error } = await supabase
            .from('conversations')
            .insert({
                user_id: user.id,
                is_active: true,
                messages: [],
            })
            .select()
            .single();

        if (error) throw error;
        return data;
    },

    /** Add a message to an existing conversation */
    addMessage: async (
        conversationId: string,
        message: { role: 'user' | 'assistant'; content: string; metadata?: any },
    ): Promise<Conversation> => {
        // First, get the current messages
        const { data: current, error: fetchError } = await supabase
            .from('conversations')
            .select('messages')
            .eq('id', conversationId)
            .single();

        if (fetchError) throw fetchError;

        const updatedMessages = [
            ...(current?.messages || []),
            { ...message, timestamp: new Date().toISOString() },
        ];

        const { data, error } = await supabase
            .from('conversations')
            .update({
                messages: updatedMessages,
                last_message_at: new Date().toISOString(),
            })
            .eq('id', conversationId)
            .select()
            .single();

        if (error) throw error;
        return data;
    },

    /** Deactivate a conversation */
    deactivate: async (conversationId: string): Promise<void> => {
        const { error } = await supabase
            .from('conversations')
            .update({ is_active: false })
            .eq('id', conversationId);

        if (error) throw error;
    },
};
